sdk/nodejs/serverless/function.ts

```diff
@@ -0,0 +1,184 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as pulumi from "@pulumi/pulumi";
+import * as cloudfunctions from "../cloudfunctions";
+import * as storage from "../storage";
+
+import * as filepath from "path";
+import * as readPackageJson from "read-package-json";
+
+/**
+ * Handler is the signature for a serverless function.
+ */
+export type Handler = (request: any, response: any) => any;
+
+/**
+ * FunctionOptions provides configuration options for the serverless Function.
+ */
+export interface FunctionOptions {
+    /**
+     * Memory (in MB), available to the function. Default value is 256MB. Allowed values are: 128MB, 256MB, 512MB, 1024MB, and 2048MB.
+     */
+    readonly availableMemoryMb?: pulumi.Input<number>;
+    /**
+     * Description of the function.
+     */
+    readonly description?: pulumi.Input<string>;
+    /**
+     * A set of key/value label pairs to assign to the function.
+     */
+    readonly labels?: pulumi.Input<{
+        [key: string]: any;
+    }>;
+    /**
+     * Project of the function. If it is not provided, the provider project is used.
+     */
+    readonly project?: pulumi.Input<string>;
+    /**
+     * Region of function. Currently can be only "us-central1". If it is not provided, the provider region is used.
+     */
+    readonly region?: pulumi.Input<string>;
+    /**
+     * Timeout (in seconds) for the function. Default value is 60 seconds. Cannot be more than 540 seconds.
+     */
+    readonly timeout?: pulumi.Input<number>;        
+   /**
+    * The packages relative to the program folder to not include the Function upload. This can be
+    * used to override the default serialization logic that includes all packages referenced by
+    * project.json (except @pulumi packages).  Default is `[]`.
+    */
+   excludePackages?: string[];
+}
+
+/**
+ * Function is a higher-level API for creating and managing GCP Cloud Function resources implemented
+ * by a given handler.
+ */
+export class Function extends pulumi.ComponentResource {
+    public readonly options: FunctionOptions;
+    public readonly function: cloudfunctions.Function;
+    public readonly bucket: storage.Bucket;
+    public readonly bucketObject: storage.BucketObject;
+
+    constructor(name: string,
+        options: FunctionOptions,
+        func: Handler,
+        opts?: pulumi.ResourceOptions,
+        serialize?: (obj: any) => boolean) {
```

Comment:
We shouldn't need this `serialize` any more in the AWS version of this function, and I'd suggest we leave it out here since there is no legacy requirement.